### ISSUE TITLE
(NU5) Witnessed note has unexpected commitment value

### ISSUE THREAD

**Issue by ebfull** [zcash/zcash#5816]
Context: I had a node on the private testnet branch syncing from a fresh wallet, which I then enabled mining after syncing with the network. I did some transactions and then attempted to spend all of my orchard notes at once, when...

This assertion is failing

```rust
assert_eq!(
    self.witness_tree
        .get_witnessed_leaf(*position)
        .expect("tree has witnessed the leaf for this note."),
    &MerkleHashOrchard::from_cmx(&dnote.note.commitment().into()),
);
```

in `Wallet::get_spend_info`. The assertion is

```
thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
  left: `MerkleHashOrchard(0x115664f313d0b72d147eef7e856ec7f6fc80b31753ba06261005172f8c0acdae)`,
 right: `MerkleHashOrchard(0x3eb0e71821bdab197cdc1e519bb6e8a30d636aefb877590bf61987a4d1d6203b)`', src/rust/src/wallet.rs:692:9
```

Both of these hashes are apparently note commitments for stuff I mined to shielded coinbase. I looked through my wallet and from what I can tell `0x115664f313d0b72d147eef7e856ec7f6fc80b31753ba06261005172f8c0acdae` was a shielded coinbase output from txid `e5c278aaea228c669a141ef2ddc736e60089731fdb26935ce7216b142726d6e3` (height 1820085) and `0x3eb0e71821bdab197cdc1e519bb6e8a30d636aefb877590bf61987a4d1d6203b` was a shielded coinbase output from txid `8aaa355fa501cd359198024baa7249e4cd5c0bcb09d73cdc02de825a41aad108` (height 1820079).

Looking in my logs I had mined a couple blocks in between these two blocks, so it's not clear to me why the wallet would be confusing the two notes with each other here. There also doesn't appear to have been any reorgs happening around that time on the network.

**Comment by ebfull** [comment]
When I booted my node up next, I got

```
2022-04-04T16:09:57.530705Z ERROR Init: rustzcash::wallet: Failed to read Orchard note commitment or last checkpoint height: Consistency violation found when attempting to deserialize Merkle tree: InvalidWitnessIndex
```

It then proceeded to output a bunch of `AddToWallet` logs _starting with_ txid `b994f9ec1f9bb4c01a20960097d76a44d1f33544b6b5cd163e15f1254a5c2063` which is a transaction I had sent earlier which sent a bunch of orchard notes I had mined to a single note. I shut my node down gracefully after it synced again.

I loaded my node again, this time doing `-rescan` and then immediately attempted to spend all my orchard notes again. The transaction failed with:

```
2022-04-04T17:15:21.377319Z ERROR rustzcash::wallet: Error obtaining spend info for outpoint OutPoint { txid: TxId([99, 32, 92, 74, 37, 241, 21, 62, 22, 205, 181, 182, 68, 53, 243, 209, 68, 106, 215, 151, 0, 150, 32, 26, 192, 180, 155, 31, 236, 249, 148, 185]), action_idx: 12 }: NoteNotPositioned(OutPoint { txid: TxId([99, 32, 92, 74, 37, 241, 21, 62, 22, 205, 181, 182, 68, 53, 243, 209, 68, 106, 215, 151, 0, 150, 32, 26, 192, 180, 155, 31, 236, 249, 148, 185]), action_idx: 12 })
```

That's txid `b994f9ec1f9bb4c01a20960097d76a44d1f33544b6b5cd163e15f1254a5c2063` where I had previously sent a bunch of money to a single orchard note. Here's the output of `z_listunspent` for that tx

```
  {
    "txid": "b994f9ec1f9bb4c01a20960097d76a44d1f33544b6b5cd163e15f1254a5c2063",
    "pool": "orchard",
    "outindex": 12,
    "confirmations": 213,
    "spendable": true,
    "address": "utest13hkwnymu02e6nlap6r6v5px37a75ag5w520797t8fehap602n6jpk6gzarh8a8ffs5cvs0wut2vd9ten8yswge5n98phtxzmmmgh65vuvahn8j953x74g4kk7zhutzzaaefyr23uuyqauaacac949edyyrduyyqytafnvjjesu3ks5wv",
    "amount": 42.49989999,
    "memo": "f60000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000>
    "change": false,
    "cmx": "0cbc0a4efd52554f54b212c7c495f05c206e75fc672624dc4e0763231ab4a2d7"
  },

```

(`cmx` appears there because I edited my zcashd to print that out for debugging purposes)

**Comment by ebfull** [comment]
Okay, the issue I identified in my second comment on this issue was caused by a bug in the Orchard rescan logic which should be addressed by #5819.

**Comment by nuttycom** [comment]
Wrote this addition to the mining_shielded_coinbase test to try to reproduce the issue: https://github.com/zcash/zcash/pull/5821

This doesn't reproduce the issue, but it *does* produce an outright crash with no trace information.

**Comment by ebfull** [comment]
It is incredibly likely that this was caused by the bug fixed in https://github.com/zcash/incrementalmerkletree/pull/35.

**Comment by steven-ecc** [comment]
Closing as this appears fixed by [zcash/incrementalmerkletree#35](https://github.com/zcash/incrementalmerkletree/pull/35).
